test/test_serialization.py

```diff
@@ -4097,6 +4098,54 @@ def test_empty_class_serialization(self):
             f.seek(0)
             tensor2 = torch.load(f)
 
+    def test_allowlist_for_weights_only(self):
+        # Doing the import here as this test later dels the import from sys.modules
+        from torch.testing._internal.two_tensor import TwoTensor
+        t = TwoTensor(torch.randn(2, 3), torch.randn(2, 3))
+        p = torch.nn.Parameter(t)
+        sd = OrderedDict([('t', t), ('p', p)])
+
+        with tempfile.NamedTemporaryFile() as f:
+            torch.save(sd, f)
+            # unimport TwoTensor
+            del sys.modules['torch.testing._internal.two_tensor']
+
+            # Loading tensor subclass with weights_only=True should fail
+            # if tensor subclass has not been imported
+            with self.assertRaisesRegex(pickle.UnpicklingError,
+                                        "`torch.testing._internal.two_tensor` was not found in `sys.modules`"):
+                f.seek(0)
+                sd = torch.load(f, weights_only=True)
+
+            # Loading tensor subclass with weights_only=True should work
+            # if __setstate__ and tp_alloc are not overriden and user has imported the subclass
+            from torch.testing._internal.two_tensor import TwoTensor
+            f.seek(0)
+            sd = torch.load(f, weights_only=True)
```

Comment:
IMO this is wrong: weights_only should refuse to deserialize user imported classes unless they are allowlisted. Because almost every Python program program imports `sys`, which makes it "safe" (which is clearly unsafe) to call functions from that module)

Reply:
Hm, the implementation in the unpickler will do this check though, so arbitrary functions from sys should not be callable (unless they are tensor subclasses that do not override `__setstate__` or `tp_alloc`

```
class_type = getattr(modules[module], name)
if isinstance(class_type, type) and issubclass(
    class_type, torch.Tensor
):
    # checks __setstate__ and tp_alloc
else:
     # raise Unsupported
```

Do you think this is still unsafe?

The rationale for this was that @albanD was hoping that we could "blanket" allow tensor subclasses that fall under certain restrictions

Reply:
I verified that other methods are not called by adding a test that overrides every single method of a class and noticed  that `__setattr__` and `__getattribute__` are called as well (due to `getattr` and `setattr` calls from `rebuild_from_type_v2` and the call to `_set_obj_state` in `rebuild_from_type_v2`.

I also blocklisted other functions that could override the behavior of `getattr` and `setattr` (`__get__`, `__set__`, `__getattr__`)